### ISSUE TITLE
regression: SpecialKey highlight overrules matchadd()/:match

### ISSUE THREAD

**Issue by ninrod** [neovim/neovim#4946]
- `nvim --version`:NVIM v0.1.5-966-ga2ecbc2
- Vim (version: ) behaves differently? yes
- Operating system/version: MacOSx yosemite
- Terminal name/version: iTerm2 3.0.2
- `$TERM`: xterm-256color
### Actual behaviour

guifg is not respected
### Expected behaviour

guifg is respected
### Steps to reproduce using `nvim -u NORC`

I don't have the knowledge to reproduce using nvim -u NORC, but I know for a fact that this block of code:

``` viml
highlight ExtraWhitespace ctermfg=166 guifg=#d65d0e 
match ExtraWhitespace /\s\+$/
autocmd BufWinEnter * match ExtraWhitespace /\s\+$/
autocmd InsertEnter * match ExtraWhitespace /\s\+\%#\@<!$/
autocmd InsertLeave * match ExtraWhitespace /\s\+$/
autocmd BufWinLeave * call clearmatches()
```

works in neovim 0.1.4 and does not work in neovim 0.1.5 (HEAD commit in this moment), both with truecolor support activated. And yes, I've correctly used the `set termguicolors` in nvim 0.1.5. for nvim 0.1.4 I used `let $NVIM_TUI_ENABLE_TRUE_COLOR=1`.

I also verified this: while `guifg` is ignored, `guibg` is correctly handled in nvim 0.1.5 with truecolor support enabled.

**Comment by justinmk** [comment]
@ninrod Could you revert https://github.com/neovim/neovim/commit/1e93e24f5e6e36fa89ebbc234b8c91a21ef0b788 and see if the problem still occurs?

**Comment by ninrod** [comment]
@justinmk will do as soon as I arrive home. Here at work I only have access to vim because I don't know how to build neovim behind a corporate proxy.

**Comment by romgrk** [comment]
Can't reproduce here. How do you test that `guifg` is respected? By definition a space character has no foreground. On my setup the cursor is reversed though, and having it hover a trailling whitespace does show up an orange bg.

**Comment by ninrod** [comment]
I don't know exactly if you're talking about this: `set list listchars=tab:\ \ ,trail:·`. But I have this setup. 

here is my [full vimrc](https://github.com/ninrod/dotfiles/blob/master/dot/vimrc)

here's the code block that will highlight for you an extrawhitespace:

```viml
highlight ExtraWhitespace ctermfg=166 guifg=#d65d0e 
match ExtraWhitespace /\s\+$/
autocmd BufWinEnter * match ExtraWhitespace /\s\+$/
autocmd InsertEnter * match ExtraWhitespace /\s\+\%#\@<!$/
autocmd InsertLeave * match ExtraWhitespace /\s\+$/
autocmd BufWinLeave * call clearmatches()
```

if you have truecolor support properly setup, in both versions, you'll be able to see that in neovim 0.1.4 guifg does work and in 0.1.5 it does not work. 

For reference, I'm using the gruvbox colorscheme with truecolor support in iTerm2.

**Comment by romgrk** [comment]
Oh, I think there were some commits touching the highlight precedence of trailing spaces. I'll check if I can find them. Btw, is `guifg` not respected in general, or just for this specific case? What does the `hi ExtraWhitespace` command outputs? What does it output for other non-related highlight groups?

**Comment by justinmk** [comment]
@ninrod Does it work correctly with true color _disabled_?

**Comment by ninrod** [comment]
@justinmk if I understand correctly, with truecolor disabled neither guifg or guibg are respected because you are placed in 256 colorspace anyway. so you have to use `ctermfg` instead of guifg. Well I think it will work, but it defeats the purpose, yes? 

and gruvbox does support 256 colors. 

I think gruvbox works well in 0.1.5, in truecolor mode, generally speaking. I think the problem is specifically confined in this code block i just posted, as @romgrk hinted.

I'll look more into it later when I can get my hands on neovim.

**Comment by justinmk** [comment]
> with truecolor disabled neither guifg or guibg are respected 

Yes, could be useful to confirm this only affects guifg, not ctermfg.

**Comment by ninrod** [comment]
@justinmk I see you point. I'll also confirm if `ctermfg` is affected or not.

**Comment by ninrod** [comment]
@justinmk I've tested the 2 cases you asked for:

0. I confirmed that commit 1e93e24 **has** the bug.
0. I confirmed that ctermfg **is not respected either**. ctermfg is affected as much as guifg is affected. In both commits: `HEAD` and 1e93e24.

So I changed the title of the issue report to reflect these new findings.

**Comment by ninrod** [comment]
@romgrk 
>  What does the hi ExtraWhitespace command outputs? 

* this:
```txt
:hi ExtraWhitespace
ExtraWhitespace xxx ctermfg=166 guifg=#d65d0e
Press ENTER or type command to continue
```

>What does it output for other non-related highlight groups?

* here
```txt
:hi GruvboxAqua
GruvboxAqua    xxx ctermfg=108 guifg=#8ec07c
Press ENTER or type command to continue
```

I've just turned on truecolor support from `vim` (HEAD commit) and it handles the ExtraWhitespace code block flawlessly. So it does behave differently from `nvim`.

**Comment by ninrod** [comment]
here is a revealing screenshot of the bug. check out lines 209 and 210.

left is nvim in commit 1e93e24
right is nvim --version 0.1.4


`nvim` in `HEAD` commit has the same behaviour.


<img width="1280" alt="bug" src="https://cloud.githubusercontent.com/assets/8352747/16307891/50021b60-3939-11e6-922d-3d94941fcf00.png">

**Comment by romgrk** [comment]
What about without this https://github.com/neovim/neovim/pull/4744?

**Comment by ninrod** [comment]
@romgrk 

I just confirmed your suspicion. This merged pull is indeed the culprit. Here are the conclusions:

66adc2d  (and up to HEAD) has the bug. (this is pull #4744, merged in master by @justinmk )
56cb1eb does **not** have the bug.

proof:

<img width="1280" alt="bughunt" src="https://cloud.githubusercontent.com/assets/8352747/16356552/92901480-3ab2-11e6-99fd-47aaed90032a.png">

**Comment by ninrod** [comment]
So I've checked out master (commit 47a15d0) and issued:
    `$ git revert -m 1 66adc2d`

This fixes the bug as shown here:

<img width="1280" alt="bugfix" src="https://cloud.githubusercontent.com/assets/8352747/16356870/c96863fe-3abc-11e6-8f54-8988dbd52a73.png">

**Comment by ninrod** [comment]
@justinmk, revisiting your suggestion that I did not understand at the time:

> @ninrod Could you revert 1e93e24 and see if the problem still occurs?

Yes, now I have checked out master again (commit 47a15d) and reverted the changes brought by merge commit 1e93e24 with `git revert -m 1 1e93e24`. the problem still occurs.

So the real fix to the issue is really to revert pull #4744 with:

`git revert -m 1 66adc2d`

**Comment by justinmk** [comment]
@zhou13 would you be able to look into this?

**Comment by ninrod** [comment]
just pouring in more information: if I change the block to only:

`highlight SpecialKey ctermfg=166 guifg=#d65d0e`

it works. 

So it must be some issue with SpecialKey precedence versus user defined highlights.

**Comment by zhou13** [comment]
I will have a look at this problem.  I saw you refer to `SpecialKey` but it is still not clear to me why it matters.  Do you enable `listchars` in you vimrc?  Could you give a minimal nvimrc and a test file to reprocedure the problem?

**Comment by ninrod** [comment]
@zhou13, if you look at vim help for SpecialKey, you'll see that it affects listchars.

I have this set up in my vimrc.

```
set list listchars=tab:\ \ ,trail:·
```
here's vim help: 
```
							*hl-SpecialKey*
SpecialKey	Meta and special keys listed with ":map", also for text used
		        to show unprintable characters in the text, 'listchars'.
	         	Generally: text that is displayed differently from what it
		        really is.
```

I can try to provide a minimal setup. Just don't know how to do that yet.

**Comment by zhou13** [comment]
I see your problem.  You want to show trailing space with `·` but you don't like the its color (specified by `SpecialKey` by default) so you tried to define you own custom highlight to "workaround" this problem.  Unfortunately, my pull #4744 increase the priority of `listchar` so your workaround does not work anymore.  It is arguable that custom highlight should overwrite the highlight of `listchar` but I can also give the cases why this is not desirable: You still want to highlight the tab and trailing space under some custom highlight.

You can temporarily workaround this problem by executing `hi clear SpecialKey` but it will give you many unwanted side effects.  For long term, I think the best way to solve this problem is to split the color `SpecialKey` because people want to set different highlights to
* trailing space in `listchar`
* tab in `listchar`
* Meta and special keys listed with ":map"
* Text used to show unprintable characters in the text

**Comment by ninrod** [comment]
I cannot see an instance when a user would want to increase listchars priority over his custom highlight rule. Can you provide an example? 

If I understand correctly, custom highlight overriding listchar is the rule while listchar overriding custom user highlight is the the exception. 

So I think that your pull really should provide an option to activate the behaviour of increasing the priority of listchar. If the option is not present, then the behaviour is the default as it always have been. 

I think that's a reasonable compromise.

**Comment by justinmk** [comment]
New options should be avoided. Let's decide on a reasonable default.

**Comment by ninrod** [comment]
I suspect that if this problem affected this specific setting of mine, it may as well affect other settings of other people. Maybe those settings are in place for years, and suddenly, when people upgrade for neovim 0.1.5, colors are all off. 

This setting, for instance, for coloring extrawhitespaces, is extremely useful to me. It's only a highlight, but it enables quick and comfortable way to identity those pesky extra whitespaces. It's a very important part of my setup. 

Is it possible to solve the problem that @zhou13's pull tries to solve without resorting to increase the highlight priority of listchars above user's custom highlight rules?

**Comment by ninrod** [comment]
Now let me take a step back. maybe I'm wrong? Maybe `SpecialKey` should indeed have a higher priority over custom user rules because if a rule covers a special char covered by listchars then, that rule should really target specialkey?

I don't think I even know what should happen in this case. Maybe #3670 is really a bug and pull #4744 is the answer?

anyway, this solves my problem in all versions, vim or nvim:

```viml
highlight SpecialKey ctermfg=166 guifg=#d65d0e 
```

plus, in HEAD, #3670 is solved. 

In other words, maybe this should not really work:

```viml
highlight ExtraWhitespace ctermfg=166 guifg=#d65d0e 
match ExtraWhitespace /\s\+$/
autocmd BufWinEnter * match ExtraWhitespace /\s\+$/
autocmd InsertEnter * match ExtraWhitespace /\s\+\%#\@<!$/
autocmd InsertLeave * match ExtraWhitespace /\s\+$/
autocmd BufWinLeave * call clearmatches()
```

because you'd had to use SpecialKey instead?

I'm lost.

**Comment by zhou13** [comment]
@ninrod If you want to set the color of trailing whitespace, it is intended to set the highlight of `SpecialKey` rather than double highlighting it using a custom rule, which causes priority problem. However, the problem here is that `SpecialKey` does not only apply to the highlight of trailing whitespace but also to the other highlight such as the special char such as `<C-U>` in the output of `:map` and vimrc.  Your
````
highlight SpecialKey ctermfg=166 guifg=#d65d0e
````
will make your color of you nvimrc looks differently.

**Comment by zhou13** [comment]
@ninrod Also just like what I said, if you don't want to use `SpecialKey` anymore, just using
````
hi clear SpecialKey
````
and your custom coloring rule will override `SpecialKey` I think.

**Comment by zhou13** [comment]
@justinmk What do you think about split the highlight of `SpecialKey` into something like `Trailing`, `Tab`, `SpecialKey`, and `Unprintable` and maybe fallback to `SpecialKey` if others are not set?  At least it would be useful for me and it will cause less confusion such as this one.

**Comment by ninrod** [comment]
wait, I understood the `:map` part, but how this line `highlight SpecialKey ctermfg=166 guifg=#d65d0e` will make the color of my vimrc different? I didn't get this.

**Comment by zhou13** [comment]
@ninrod My mistake.  It won't affect it.

**Comment by ninrod** [comment]
np @zhou13! 

So in conclusion, I think we will only have a problem with your pull if user wants to give different colors to each unprintable char. Like, I want that red `tab` and blue trailing whitespaces? That perhaps seems minor. maybe this is not a bug at all. Is my understanding correct? 

As for your suggestion, @zhou13, I think SpecialKey should be split in 2 categories 
* `Map` for special chars in  `:map`
* `Unprintable`: for all unprintable chars. (`tab belongs here)

But really, I wonder if `:map` is so important to deserve it's own highlight group. If user correctly configured his SpecialKey colouring, then `:map` should print an acceptable output already. I cannot see the need to implement this split. 

Maybe we should close this an document this behaviour in neovim help? even add a changelog entry?

Cannot see this as a bug anymore.

**Comment by justinmk** [comment]
Is there anything left to do here?

> @justinmk What do you think about splitting the highlight of SpecialKey into something like Trailing, Tab, SpecialKey, and Unprintable. At least it would be useful for me since I don't want them to be the same color and it will cause less confusion such as this one. However, the compatibility with VIM will be an issue.

I'm not opposed to it, they could be `:hi link`'d to `SpecialKey` by default. Though it might not be necessary to be so granular: instead have only 2 groups: `SpecialKey` and e.g. `SpecialKeyUnprintable`.

**Comment by ninrod** [comment]
maybe we should include a note in the docs about the incresead priority of listchar highlighting brough by @zhou13's pull. Other than this, I think not.

**Comment by zhou13** [comment]
> As for your suggestion, @zhou13, I think SpecialKey should be split in 2 categories

> I'm not opposed to it, they could be :hi link'd to SpecialKey by default. Though it might not be necessary to be so granular: instead have only 2 groups: SpecialKey and e.g. SpecialKeyUnprintable.

@justinmk @ninrod Well, at least I want them to be in such fine granular. I like to use a low contrast color to show the tabs just like the color of my EOL symbols to distinguish them from space without being distracting.  Since the trailing spaces usually are not preferred, I'd like to highlight them with a red background color so they are noticeable.  For the highlight in `:map` and unprintable, I would like to keep them to be the default color in my current scheme for now, but someone else may want to distinguish between them.

> But really, I wonder if :map is so important to deserve it's own highlight group. If user correctly configured his SpecialKey colouring, then :map should print an acceptable output already. I cannot see the need to implement this split.

For now, I used a low contrast color for `SpecialKey` so that it won't be distracting to show the tabs, but that gives me very low readability for `:map`, which bugs me a lot though `:map` is rarely used.

If these these are OK, I can start to work on a new pull request.  For the document, I think add a new paragraph to the "3. Changed features" section in `vim_diff.txt` should be good?

**Comment by zhou13** [comment]
Furthermore, if someone wants to customize the highlight of trailing space using code
````
highlight ExtraWhitespace ctermfg=166 guifg=#d65d0e 
match ExtraWhitespace /\s\+$/
autocmd BufWinEnter * match ExtraWhitespace /\s\+$/
autocmd InsertEnter * match ExtraWhitespace /\s\+\%#\@<!$/
autocmd InsertLeave * match ExtraWhitespace /\s\+$/
autocmd BufWinLeave * call clearmatches()
````
they can just add
````
hi clear Tab  " or SpecialKeyTab
````
without affecting other color.

**Comment by justinmk** [comment]
Reviewing the thread here, and noting the documentation of `:help matchadd()` (and `:help :match`):

> Syntax highlighting (see 'syntax') is a separate mechanism, and regardless of the chosen priority **a match will always overrule syntax highlighting.**

I'm reminded that the expected behavior of `matchadd()` (and `:match`) is that it overrules normal highlighting. This is important for plugins like easymotion.

> It is arguable that custom highlight should overwrite the highlight of listchar but I can also give the cases why this is not desirable: For example, you still want to highlight the tab and trailing space under search highlight.

The search highlight example doesn't seem important. It's more important that `matchadd()` and `:match` continue to behave predictably.

> Well, at least I want them to be in such fine granular. I like to use a low contrast color to show the tabs just like the color of my EOL symbols to distinguish them from space without being distracting.

Isn't that _already_ possible using `matchadd()` with the correct pattern(s)? I'm thinking this is why the granularity isn't needed. Why can't the priority implemented in https://github.com/neovim/neovim/pull/4744 be changed to be overruled by `matchadd()`/`:match`? What would it break?

**Comment by zhou13** [comment]
You are right.

**Comment by ninrod** [comment]
See if I understand correctly: the problem here is that I'm specifying a `match` rule that is not overruling a highlight rule (SpecialKey's highlight rule) but it should. Based on this, it's really a bug.

is that right?

**Comment by justinmk** [comment]
@ninrod Right, `:match` and `matchadd()` always overrule other highlighting.

**Comment by s0undt3ch** [comment]
I recently updated to 0.1.5(Arch linux) and got hit by this problem using [this solarized](https://github.com/frankier/neovim-colors-solarized-truecolor-only) theme. 0.1.4 works.

**Comment by Shougo** [comment]
@s0undt3ch Have you set `termguicolors` option?

**Comment by s0undt3ch** [comment]
Bah! I tried once I was in nvim, setting it in `~/.config/nvim/init.vim` works

**Comment by justinmk** [comment]
@zhou13 Were you able to look into this? Reminder that we just need `matchadd()`/`:match` to be considered in the changes from  #4744.

https://github.com/neovim/neovim/issues/4946#issuecomment-228627970

**Comment by zhou13** [comment]
@justinmk  Yes.  This bug is on my TODO list.  However, I need to prepare for my prelim exam in this summer and I should be able to work on it by the middle of September.  Feel free to ping me by then.  Thanks.

**Comment by justinmk** [comment]
@zhou13 No problem, no rush. Thanks!
